AD_query_tool.py

```diff
@@ -0,0 +1,99 @@
+
+import os
+import ldap
+import csv
+import pandas as pd
+import datetime
+import dateutil.relativedelta
+
+
+def active_directory_conn(ldap_server, ldap_login, ldap_pass):
+    conn = ldap.initialize(ldap_server)
+    conn.simple_bind_s(ldap_login, ldap_pass) 
+    return(conn)
+
+def search_active_directory(base_dn, search_object,  search_attr, connection):
+    results = connection.search_s(base_dn, ldap.SCOPE_SUBTREE, search_object , search_attr)
+    connection.unbind()
+    return(results[:-4]) # cutting out irrelevant data rows 
+
+def results_to_csv(path, active_directory_query):
+    with open(path, 'w') as csv_file:  
+        writer = csv.writer(csv_file)
+        header = ['CN', 'GroupsMembership', 'UserAccountControl', 'LastLogon', 'LastPasswordSet'] 
+        writer.writerow(header)
+        for i in active_directory_query:
+            cn = i[0]
+            member_of = i[0],i[1].get('memberOf')
+            user_account_control = int(i[1].get('userAccountControl')[0])
+            last_logon = int(i[1].get('lastLogon')[0])
+            pwd_last_set = int(i[1].get('pwdLastSet')[0])
+            writer.writerow([cn, member_of ,user_account_control, last_logon, pwd_last_set])
+
+def enabled_users(df):    
+    enabled_users = df.loc[df['UserAccountControl'] & 2 == 0]
+    return(enabled_users)
+
+def enabled_users_amount():    
+    df = pd.read_csv(csv_path)
+    enabled_users_amount = enabled_users(df)['CN'].count()
+    return(enabled_users_amount)
+
+def current_date_time():
+    current_ts = datetime.datetime.now()
+    current_date_time = current_ts.replace(microsecond=0)
+    return(current_date_time)
+
+def to_date_time(timestamp):
+    date_time = datetime.datetime(1601, 1, 1) + datetime.timedelta(seconds=timestamp/10000000)
+    return(date_time.replace(microsecond=0))
+
+def password_age():
+    df = pd.read_csv(csv_path)
+    password_age_list = []
+    for timestamp in df['LastPasswordSet']:
```

Comment:
I'm really not sure, but can this loop be done with pandas? like adding a column with the deltas

Reply:
Pandas can create a column as an output of some calculations, but in this case, the calculation is a delta between a single value and a series (a column). So it can be done, but I think that it's more complicated than a simple loop (for example https://stackoverflow.com/questions/35327450/using-datetime-timedelta-with-a-series-in-a-pandas-df/35327542)